### ISSUE TITLE
Memory leak when moving cursor across nested contexts

### ISSUE THREAD

**Issue by Ottatop** [nvim-treesitter/nvim-treesitter-context#296]
### Description

Moving the cursor across a nested context such that a context line is added or removed will leak some memory. This doesn't happen when moving the cursor between no context and the first level of context. Navigating to other files won't release leaked memory, but disabling the plugin with `:TSContextDisable` will.

I noticed that after a few hours programming with my nvim config, memory usage would be several gigabytes large. I narrowed this down to scrolling up and down and eventually to nvim-treesitter-context. While the amount of leakage in the minimal example isn't anywhere near as high as the leakage in my own config, it's still noteworthy.

### Neovim version

0.9.1

### Expected behavior

Memory usage doesn't increase forever.

### Actual behavior

Memory usage does increase, and won't be freed until the plugin is disabled or nvim is exited.

### Minimal config

```Lua
local plugins = {
  ts         = 'https://github.com/nvim-treesitter/nvim-treesitter',
  ts_context = 'https://github.com/nvim-treesitter/nvim-treesitter-context',
  -- ADD ADDITIONAL PLUGINS THAT ARE _NECESSARY_ TO REPRODUCE THE ISSUE
}

for name, url in pairs(plugins) do
  local install_path = '/tmp/nvim/site/'..name
  if vim.fn.isdirectory(install_path) == 0 then
    vim.fn.system { 'git', 'clone', '--depth=1', url, install_path }
  end
  vim.o.runtimepath = install_path..','..vim.o.runtimepath
end

-- ADD INIT.LUA SETTINGS THAT IS _NECESSARY_ FOR REPRODUCING THE ISSUE
require("nvim-treesitter.configs").setup({
  auto_install = true,
})
```

### Steps to reproduce

1. `nvim --clean -u minimal.lua`
2. Navigate to a file with nested contexts.
3. Repeatedly move the cursor up and down between two nested contexts. Memory usage should begin to increase.
4. Move the cursor up and down between no context and some context. Memory usage doesn't increase.
5. Do `:TSContextDisable`. This will free the leaked memory.

**Comment by lewis6991** [comment]
This might not have fixed the problem entirely, but 4d497f8e5618cc5dd225d028025c76cfd93fb78d should reduce any memory usage.

**Comment by Ottatop** [comment]
Yep, that seems to have fixed it!
